make command refactor (#109)
aerogear/mobile-security-service-operator: README.adoc

$ minishift start
----

=== Preparing the Project

Use the following command to setup the cluster role, service accounts, monitoring and CRDS required for the project

[source,shell]
----

> $ make setup-project

Reviewer: ```suggestion
$ make install
```

----

> === Installing Mobile Security Service by Operator

Reviewer: ```suggestion
IMPORTANT: To installl, you need to be logged with a user with full access in the cluster.  E.g. Use `oc login -u system:admin`.

=== Installing Mobile Security Service by Operator
```

Use the following command to create the Operator, Service and its Database.

[source,shell]
----

> $ make install-mss

Reviewer: As described above it shows more complexy for me and in POV no makes sense add more steps. 
Also, the nomenclature install-mss for me means install just the Service

+
[source,shell]
----
$ make create-example-app
----

NOTE: This command will just apply the link:./deploy/crds/examples/mobile-security-service_v1alpha1_mobilesecurityserviceapp_cr.yaml[MobileSecurityServiceApp CR] example with `kubectl apply -f deploy/crds/examples/mobile-security-service_v1alpha1_mobilesecurityserviceapp_cr.yaml`

=== Deleting/Unbinding an app

* Run the following command to delete/unbind the app from the service and remove its SDKConfigMap.
+
[source,shell]
----
$ make delete-example-app
----

IMPORTANT: This command will just delete the link:./deploy/crds/examples/mobile-security-service_v1alpha1_mobilesecurityserviceapp_cr.yaml[MobileSecurityServiceApp CR] example with `kubectl delete -f deploy/crds/examples/mobile-security-service_v1alpha1_mobilesecurityserviceapp_cr.yaml`
NOTE: The Rest Service endpoint to delete it is called in the finalizer of the link:./deploy/crds/examples/mobile-security-service_v1alpha1_mobilesecurityserviceapp_cr.yaml[MobileSecurityServiceApp CR] and the CR will be just allowed to be removed when the app is no longer available in its Service.

> === Uninstalling the Application and Operator

Reviewer: ```suggestion
=== Uninstalling
```

Use the following command to delete the Operator, the Service and its Database and all related configuration for the operator.

[source,shell]
----

> $ make clean

Reviewer: Was not the name changed to uninstall? Has places that it still with the clean one.

----

== Configuration and Options

=== Operator Namespace

> By using the make command `make install` the default namespace, e.g `mobile-security-service`, will be created and the operator will be installed in it. You are able to install the operator in another namespace as you wish, however, you need setup its roles (RBAC) to by applied on the namespace where the operator will be installed. The namespace name need to be changed in the link:./deploy/cluster_role_binding.yaml[Cluster Role Binding] file in the following line.

Reviewer: ```suggestion
By using the make command `make install` the default namespace, e.g `mobile-security-service`,  defined in the link:./Makefile[Makefile] will be created and the operator will be installed in it. You are able to install the operator in another namespace as you wish, however, you need to set up its roles (RBAC) in order to apply them on the namespace where the operator will be installed. The namespace name need to be changed in the link:./deploy/cluster_role_binding.yaml[Cluster Role Binding] file in the following line. Note that you also can change the namespace in the link:./Makefile[Makefile] in order to still using the command `make install` for another namespace. 
```

+
[source,shell]
----

> $ make bind-app

Reviewer: IHMO we should not use the nomenclature bind since it is not a bind. What does? It applies an APP CR example, so it should be apply-app-example.

```suggestion
$ make apply-app-example
```

+
[source,shell]
----

> $ make unbind-app

Reviewer: I do not think we should use unbind. What does it? It will delete the APP CR example. 
```suggestion
$ make delete-app-example
```

----

IMPORTANT: This command will just delete the link:./deploy/crds/examples/mobile-security-service_v1alpha1_mobilesecurityserviceapp_cr.yaml[MobileSecurityServiceApp CR] example with `kubectl delete -f deploy/crds/examples/mobile-security-service_v1alpha1_mobilesecurityserviceapp_cr.yaml`
NOTE: The Rest Service endpoint to delete it is called in the finalizer of the link:./deploy/crds/examples/mobile-security-service_v1alpha1_mobilesecurityserviceapp_cr.yaml[MobileSecurityServiceApp CR] and the CR will be just allowed to be removed when the app is no longer available in its Service.

> === Removing the Application

Reviewer: ```suggestion
=== Uninstall the Operator
```

The application-monitoring stack provisioned by the
https://github.com/integr8ly/application-monitoring-operator[application-monitoring-operator]
can be used to gather metrics from the operator here.  Once you have

> provisioned that (or the ServiceMonitor CRD at a minimum), it is configured by default with `make setup-project` 

Reviewer: The monitor service should not be installed by default. It is just valid for integr8ly. See :  https://github.com/aerogear/mobile-security-service-operator/pull/106

$ kubectl create -f deploy/monitor/grafana-dashboard.yaml
----

> This operator will just working with the namespaces which are specified in the environment variable `APP_NAMESPACE` and the link:./deploy/crds/examples/mobile-security-service_v1alpha1_mobilesecurityserviceapp_cr.yaml[MobileSecurityServiceApp CR] which be applied in a NAMESPACE which was not defined on it will be ignored. See its configuration into the link:./deploy/operator.yaml[operator.yaml] file.

Reviewer: ```suggestion
This operator will just working with the namespaces which are specified in the environment variable `APP_NAMESPACE` and the link:./deploy/crds/examples/mobile-security-service_v1alpha1_mobilesecurityserviceapp_cr.yaml[MobileSecurityServiceApp CR] which be applied in a NAMESPACE which was not defined on it will be ignored. See the EnvVar configuration into the link:./deploy/operator.yaml[operator.yaml] file. Also, note that the valid namespaces can be managed by the OCP UI. 
```

$ kubectl create -f deploy/monitor/grafana-dashboard.yaml
----

> This operator will only working with the namespaces which are specified in the environment variable `APP_NAMESPACE`. Any link:./deploy/crds/examples/mobile-security-service_v1alpha1_mobilesecurityserviceapp_cr.yaml[MobileSecurityServiceApp CR] which is created/applied in a NAMESPACE which is not defined in `APP_NAMESPACE` will be ignored. See the EnvVar configuration into the link:./deploy/operator.yaml[operator.yaml] file. Also, note that the valid namespaces can be managed by the OCP UI. 

Reviewer: The operator do not work win the APP_NAMESPACE.

Reviewer: The  Namespace to apply MobileSecurityServiceApp has the info regarsd this configuration.

> This operator will only working with the namespaces which are specified in the environment variable `APP_NAMESPACES`. Any link:./deploy/crds/examples/mobile-security-service_v1alpha1_mobilesecurityserviceapp_cr.yaml[MobileSecurityServiceApp CR] which is created/applied in a NAMESPACE which is not defined in `APP_NAMESPACES` will be ignored. See the EnvVar configuration into the link:./deploy/operator.yaml[operator.yaml] file. Also, note that the valid namespaces can be managed by the OCP UI.

Reviewer: This section is `Monitoring Service (Metrics)` and it just over it, in this way this information shows not accurate here. See that w have a section i the README which explains it. https://github.com/aerogear/mobile-security-service-operator#namespace-to-apply-mobilesecurityserviceapp

PS.: It `This operator will only working with the namespaces which are specified in the environment variable `APP_NAMESPACES`` is not true. The operator also working on with the namespace where it was installed/applied. 

c/c @davidffrench @laurafitzgerald

Author: There was feedback about this item provided here: https://github.com/aerogear/mobile-security-service-operator/pull/109#discussion_r291562686 I don't believe that I originally added this as part of this change as the README changes should have been limited to only changes which were required to reflect the Makefile. I am happy to have this removed from here if it's not appropriate having it in this section or updated. Can you open a pr for that @camilamacedo86

Reviewer: Hi @laurafitzgerald the comment  #109 (comment) was regards a typo. I think it was added but it is fine we can do another PR no problem at all. I just would like to share the knowledge with you.

----

NOTE:  `make uninstall-monitoring` will uninstall the Service.

> NOTE: To run the project locally export the ENV VAR. E.g `export APP_NAMESPACE=mobile-security-service-apps`

Reviewer: ```suggestion

```

== Makefile command reference

> === Application Setup/Install/Teardown

Reviewer: Teardown for me is not the usual nomenclature used for that. (It is usually used into integration tests)
The system/tools usually have the commands install and uninstall.

|===
| *Command*                        | *Description*
| `make setup-project`             | Create {namespace} namespace, monitoring, application CRDS, cluster role and service account
| `make teardown`                  | Delete {namespace} namespace, monitoring, application CRDS, cluster role and service account
| `make install-mss`               | Installs the MSS Operator and the MSS CRs.
| `make uninstall-mss`             | Uninstalls the MSS Operator and the MSS CRs.
| `make bind-app`                  | Apply the App CR . (Create/Update app in the cluster and Service, also creates ConfigMap with the public host endpoint for the init config in the mobile device(SDK) ).
| `make ubind-app`                 | Delete the App CR. (Delete app from the Service and SDKConfigMap).
| `make install-service-and-db`     | Create Mobile Security Service App and its database in the project.
| `make uninstall-service-and-db`     | Delete Mobile Security Service App and its database 
|===

> === Container Build and Push

Reviewer: ```suggestion
=== Commands used by the CI
```

|===
| `make build-dev`                 | Build operator dev image with tag `quay.io/aerogear/mobile-security-service-operator:<version>-dev`.
| `make push-dev`                  | Push operator dev image to https://quay.io/repository/aerogear/mobile-security-service-operator[quay.io].

> | `make build-master`              | Used by CI to build operator image from `master` branch and add `:master` tag.

Reviewer: ```suggestion
| `make build-master`              | Used by build operator image from `master` branch and push to the `:master` tag.
```

> | `make push-release`              | Used by CI to push image built by `make build-release` to https://quay.io/repository/aerogear/mobile-security-service-operator[quay.io registry].

Reviewer: ```suggestion
| `make push-release`              | Used to push image built by `make build-release` to https://quay.io/repository/aerogear/mobile-security-service-operator[quay.io registry].
```

> | `make build-latest`              | Used by CI to build operator image from a tagged commit and add `:latest` tag.

Reviewer: ```suggestion
| `make build-latest`              | Used to build operator image from a tagged commit and push to `:latest` tag.
```

> | `make push-latest`               | Used by CI to push image built by `make build-latest` to https://quay.io/repository/aerogear/mobile-security-service-operator[quay.io registry].

Reviewer: ```suggestion
| `make push-latest`               | Used to push image built by `make build-latest` to https://quay.io/repository/aerogear/mobile-security-service-operator[quay.io registry].
```

|===

> === Run/Format/Test

Reviewer: I liked it 👍

|===
| `make run-local`                 | Run the operator locally for development purposes.
| `make debug-setup`               | Setup environment for debug proposes.
| `make vet`                       | Examines source code and reports suspicious constructs using https://golang.org/cmd/vet/[vet].
| `make fmt`                       | Formats code using https://golang.org/cmd/gofmt/[gofmt].

> | `make test`                      | Run test suite

Reviewer: Add it to the CI responsibility table.

> | `make test-integration-cover`    | Run coverage check

Reviewer: Add it to the CI responsibility table.

== Makefile command reference

=== Application Commands

|===
| *Command*                        | *Description*

> | `make install`                   | Create {namespace} namespace, application CRDS, cluster role and service account. Install the operator and the Service and DB

Reviewer: ```suggestion
| `make install`                   | Create the default namespace and apply operator, service and database as the configuration files on it.
```

> | `make uninstall`                 | Uninstall the operator and the Service and DB. Delete {namespace} namespace, application CRDS, cluster role and service account.

Reviewer: ```suggestion
| `make uninstall`                 | Uninstall all configuration performed by the `make install` command
```

| `make apply-example-app`         | Apply the Example App CR . (Create/Update app in the cluster and Service, also creates ConfigMap with the public host endpoint for the init config in the mobile device(SDK) ).
| `make delete-example-app`        | Delete the Example App CR. (Delete app from the Service and SDKConfigMap).
| `make refresh-operator-image`    | Deletes and applies the operator in order to refresh the image when a tag is not changed (development use)
| `make install-monitoring`        | Install Monitoring Service in order to provide metrics

> | `make uninstall-monitoring`      | Uninstall Monitoring Service in order to provide metrics

Reviewer: ```suggestion
| `make uninstall-monitoring`      | Uninstall all configuration performed by the `make install-monitoring` command
```

|===

=== Local Development

|===

> | `make run-local`                 | Run the operator locally for development purposes.

Reviewer: ```suggestion
| `make run-local`                 | Run the operator locally for development purpose. (`operator-sdk up local`)
```

> | `make debug-setup`               | Setup environment for debug proposes.

Reviewer: ```suggestion
| `make debug-setup`               | Setup environment for debugging purpose.
```

> | `make install-monitoring`        | Install Monitoring Service in order to provide metrics

Reviewer: I think it is in the application section. It is not specific for local dev

> | `make uninstall-monitoring`      | Uninstall Monitoring Service in order to provide metrics

Reviewer: The same. I think it is in the application section. It is not specific for local dev

== Makefile command reference

=== Application Commands

|===
| *Command*                        | *Description*

> | `application/install`                   | Creates the `{namespace}` namespace, application CRDS, cluster role and service account. Installs the operator and the Service and DB

Reviewer: Why it was changed again? Do we not agree following the standard and does as any other tool? ( install/unistall ) 
No make sense application/ since what we call as the app is the service(rest +ui) in many places.

| `make setup/debug`               | Sets up environment for debugging proposes.
| `make image/build/dev`                 | Builds operator dev image with tag `quay.io/aerogear/mobile-security-service-operator:<version>-dev`.
| `make image/push/dev`                  | Pushes operator dev image to https://quay.io/repository/aerogear/mobile-security-service-operator[quay.io].
| `make code/vet`                       | Examines source code and reports suspicious constructs using https://golang.org/cmd/vet/[vet].
| `make code/fmt`                       | Formats code using https://golang.org/cmd/gofmt/[gofmt].
|===

=== CI

|===
| `make test/run`                      | Runs test suite
| `make test/integration-cover`    | Run coverage check
| `make image/build/master`              | Used by CI to build operator image from `master` branch and add `:master` tag.
| `make image/push/master`               | Used by CI to push the `master` image to https://quay.io/repository/aerogear/mobile-security-service-operator[quay.io registry].
| `make image/build/release`             | Used by CI to build operator image from a tagged commit and add `:<version>` tag.

> | `make image/push/release`              | Used by CI to push the `release` image to https://quay.io/repository/aerogear/mobile-security-service-operator[quay.io registry].

Reviewer: I did not agree with this usage image/push/release. The common standard is "make command-command"

Reviewer 2: +1 for following convention

Reviewer 3: I am finding it hard to get a resource on the standard target naming convention. I really liked the `image/push/release` naming style as it easily split targets by category.

Reviewer 2: Yes, I should have added `but I'm not sure what the convention is`. I did find [this](https://ftp.gnu.org/old-gnu/Manuals/make-3.79.1/html_chapter/make_14.html) which provides some guidance though, fyi

== Makefile command reference

> === Application Un/Install

Reviewer: ```suggestion
=== Application 
```

Author: @camilamacedo86 I'm not sure I understand what the change request is here. I think you are looking for a word which summarises the commands underneath which are install/uninstall the application and apply/delete the app cr and refresh the operator image. I'm not sure solution is the correct one but I'm stuck for another option. Would `Application Commands` be enough? as the other ones are `Local Development` and `CI`.

Reviewer: Application Commands shows fine as well. I'd just not prefer to have the un/install since it has more commands as well.

Author: @camilamacedo86 yep makes sense. I had that thought as well. I'll go with Application Commands in that case.

|===
| *Command*                        | *Description*
| `make install`                   | Create {namespace} namespace, application CRDS, cluster role and service account. Install the operator and the Service and DB

> | `make clean`                     | Uninstall the operator and the Service and DB. Delete {namespace} namespace, application CRDS, cluster role and service account.

Reviewer: Why the namespace is not so longer deleted and why the name clean instead of uninstall which is what it is doing?
See other tools/commands E.g brew install <X> brew uninstall <X>

Author: The command does still delete the namespace and it's listed in the description. Renamed clean to uninstall so resolving.

| `make apply-example-app`         | Apply the Example App CR . (Create/Update app in the cluster and Service, also creates ConfigMap with the public host endpoint for the init config in the mobile device(SDK) ).
| `make delete-example-app`        | Delete the Example App CR. (Delete app from the Service and SDKConfigMap).

> | `make refresh-operator-image`    | Deletes and applies the operator in order to refresh the image when a tag is not changed (development use)

Reviewer: Why we need refresh one? If we would like to do it is just re-run the install command.
